src/Entity/Listener/OrderListener.php

```diff
@@ -32,4 +33,35 @@ public function preFlush(Order $order, PreFlushEventArgs $args)
             }
         }
     }
+
+    public function preUpdate(Order $order, PreUpdateEventArgs $args)
+    {
+        $objectManager = $args->getObjectManager();
+
+        if ($args->hasChangedField('number')) {
+            $number = $args->getNewValue('number');
+
+            if (null !== $number) {
+                $numberWithoutCollision = $this->findNumberWithoutCollision($objectManager, $number);
+
+                if ($numberWithoutCollision !== $number) {
+                    $args->setNewValue('number', $numberWithoutCollision);
+                }
+            }
+        }
+    }
+
+    private function findNumberWithoutCollision(EntityManagerInterface $objectManager, string $number)
+    {
+        $orderRepository = $objectManager->getRepository(Order::class);
+        $orderWithSameNumber = $orderRepository->findOneByNumber($number);
+
+        if (null === $orderWithSameNumber) {
+            return $number;
+        }
+
+        $nextId = $orderRepository->fetchNextSeqId();
+        $newNumber = Crockford::encode($nextId);
+        return $this->findNumberWithoutCollision($objectManager, $newNumber);
+    }
```

Comment:
Always use safety vest in the presence of recursive functions :safety_vest: :repeat: 

```suggestion
    private function findNumberWithoutCollision(EntityManagerInterface $objectManager, string $number, int $depth = 10)
    {
        if ($depth <= 0) {
            throw new \Exception('Could not find a unique number');
        }

        $orderRepository = $objectManager->getRepository(Order::class);
        $orderWithSameNumber = $orderRepository->findOneByNumber($number);

        if (null === $orderWithSameNumber) {
            return $number;
        }

        $nextId = $orderRepository->fetchNextSeqId();
        $newNumber = Crockford::encode($nextId);
        return $this->findNumberWithoutCollision($objectManager, $newNumber, $depth - 1);
    }
```